File: apps/downstream_jsclient/rsi.js




class Attribute
{
  constructor( name, type, size )
  {
    this.m_name = name;
    this.m_type = type;
    this.m_size = size;
    // this is the underlying binary data
    this.m_buffer = new ArrayBuffer( this.memsize() );
    // we have this here for convienience...
    switch(type)
    {
      case Attribute.EType.EUndefined:break;
      case Attribute.EType.EDouble:break;
      case Attribute.EType.EInteger:break;
      case Attribute.EType.EString:break;
      case Attribute.EType.EFloat:
      case Attribute.EType.EC3f:
      case Attribute.EType.EP3f:
      case Attribute.EType.EV3f:
      case Attribute.EType.EN3f:
      case Attribute.EType.EM44f:this.m_arrayview = new Float32Array(this.m_buffer);break;
      case Attribute.EType.EM44d:break;
      case Attribute.EType.EPtr:break;
    }; 
  }

  array()
  {
    return this.m_arrayview;
  }

  memsize()
  {
    return this.m_size*Attribute.element_size(this.m_type);
  }


  static element_size( type )
  {
    switch(type)
    {
      case this.EType.EUndefined:return -1;
      case this.EType.EFloat:return 4;
      case this.EType.EDouble:return 8;
      case this.EType.EInteger:return 4;
      case this.EType.EString:return -1;
      case this.EType.EC3f:return 4*3;
      case this.EType.EP3f:return 4*3;
      case this.EType.EV3f:return 4*3;
      case this.EType.EN3f:return 4*3;
      case this.EType.EM44f:return 4*16;
      case this.EType.EM44d:return 8*16;
      case this.EType.EPtr:return 4;
    };
    return -1;  
  }
}

Attribute.EType = 
{
    EUndefined:0,
    EFloat:1,
    EDouble:2,
    EInteger:3,
    EString:4,
    EC3f:5, // color
    EP3f:6, // point
    EV3f:7, // vector
    EN3f:8, // normal
    EM44f:9,
    EM44d:10,
    EPtr:11  
}

EOpCode =
{
    ENOP : 0,
    EMessage : 1,
    ESetAttr : 2
}

// CLIENT API functions ======================================================

// here we define the client api functions which produce and return command objects.
// Command objects are binary representation of scene edits and can be used directly
// for sending it over the wire.

function message( text )
{
  buf = new OBuffer();
  buf.write_int(EOpCode.EMessage);
  buf.write_string(text);
  return buf.to_arraybuffer();
}


function setAttr( object, attr_list )
{
  buf = new OBuffer();
  buf.write_int(EOpCode.ESetAttr);
  buf.write_string(object);

  var nattrs = attr_list.length;
  buf.write_int(nattrs);

  for( var i=0;i<nattrs;++i )
  {
    var attr = attr_list[i];
    buf.write_string(attr.m_name);
    buf.write_int(attr.m_type);
    buf.write_int(attr.m_size);
    buf.write_binary(attr.m_buffer);

  }

  return buf.to_arraybuffer();
}




// SERIALIZATION HELPERS =====================================================

// here we give very simple binary buffers for serialization and deserialization
// these are used to pack scene manipulation commands into binary messages which
// can go over the wire

class OBuffer
{
  constructor()
  {
    this.m_size = 0;
    this.m_data = new Array();
  }

  write_int( value )
  {
    this.m_data.push( {type:"int", value:value, offset:this.m_size} );
    this.m_size += 4;
  }

  write_float( value )
  {
    this.m_data.push( {type:"float", value:value, offset:this.m_size} );
    this.m_size += 4;
  }

  write_string( string )
  {
    this.m_data.push( {type:"string", value:string, offset:this.m_size} );
    this.m_size += 4 + string.length; // additional 4 bytes for size int
  }

  write_binary( arraybuffer )
  {
    this.m_data.push( {type:"binary", value:arraybuffer, offset:this.m_size} );
    this.m_size += 4 + arraybuffer.byteLength;
  }

  to_arraybuffer()
  {
    var littleEndian = true;
    var encoder = new TextEncoder();

    var buffer = new ArrayBuffer(this.m_size);
    var dataview = new DataView(buffer);

    var numItems = this.m_data.length;
    for (var i = 0; i < numItems; ++i)
    {
      var data = this.m_data[i];
      var type = data.type;

      if(type == "int")
      {
        dataview.setInt32( data.offset, data.value, littleEndian );
      }
      else
      if(type == "float")
      {
        dataview.setFloat32( data.offset, data.value, littleEndian );
      }
      else
      if(type == "string")
      {
        var size = data.value.length;
        dataview.setInt32( data.offset, size, littleEndian );
        var uint8array = encoder.encode(data.value);
        for( var j=0;j<size;++j )
          dataview.setUint8( data.offset+4+j, uint8array[j] );
      }
      else
      if(type == "binary")
      {
        var size = data.value.byteLength;
        dataview.setInt32( data.offset, size, littleEndian );

        var uint8_src = new Uint8Array(data.value);
        var uint8_dest = new Uint8Array(buffer, data.offset+4, data.value.byteLength);
        uint8_dest.set( uint8_src );
      }
    }
    return dataview.buffer;
  }
}

File: src/rsi.cpp
#include <rsi/rsi.h>
#include <stdexcept>
#include <vector>




std::string Attribute::type_str( int type )
{
	switch(type)
	{
		case EUndefined:return "undefined";
		case EFloat:return "float";
		case EDouble:return "double";
		case EInteger:return "integer";
		case EString:return "string";
		case EC3f:return "C3f";
		case EP3f:return "P3f";
		case EV3f:return "V3f";
		case EN3f:return "N3f";
		case EM44f:return "M44f";
		case EM44d:return "M44d";
		case EPtr:return "void*";
	};
	return "undefined";
}


int Attribute::element_size( int type )
{
	switch(type)
	{
		case EUndefined:return -1;
		case EFloat:return sizeof(float);
		case EDouble:return sizeof(double);
		case EInteger:return sizeof(int);
		case EString:return -1;
		case EC3f:return sizeof(float)*3;
		case EP3f:return sizeof(float)*3;
		case EV3f:return sizeof(float)*3;
		case EN3f:return sizeof(float)*3;
		case EM44f:return sizeof(float)*16;
		case EM44d:return sizeof(double)*16;
		case EPtr:return sizeof(void*);
	};
	return -1;	
}


enum EOpCode
{
	ENOP = 0,
	EMessage = 1,
	ESetAttr = 2,
	ECreate = 3,
	ERemove = 4
};

const std::string getString(const Attribute& attr, int index )
{
	int size = 0;
	const char* ptr = (char*)attr.m_data.get();
	for( int i=0;i<=index;++i )
	{
		 size = *((int*)ptr);
		 ptr += 4 + size;
	}

	ptr -= size;
	return std::string(ptr, size);
}



// SERIALIZATION HELPERS =====================================================

// here we give very simple binary buffers for serialization and deserialization
// these are used to pack scene manipulation commands into binary messages which
// can go over the wire

struct OBuffer
{
    OBuffer():message(std::stringstream::out | std::stringstream::binary)
    {
    }

    void write( const int& value )
    {
        message.write( reinterpret_cast<const char*>(&value), sizeof(int) );
    }

    void write( const float& value )
    {
        message.write( reinterpret_cast<const char*>(&value), sizeof(float) );
    }

    void write( const std::string& str )
    {
        int size = str.size();
        write( size );
        message.write( reinterpret_cast<const char*>(&str[0]), size );
    }

    void write( const char* ptr, int size )
    {
        write( size );
        message.write( ptr, size );
    }

    // returns a copy of the stream
    std::string to_string()
    {
        return message.str();
    }

 private:
    std::ostringstream message;
};


struct IBuffer
{
    IBuffer( char* data, int size ):message(std::stringstream::in | std::stringstream::binary)
    {
    	message.rdbuf()->pubsetbuf(data,size);
    }

    int read_int()
    {
    	int value = -1;
    	message.read( (char*)&value, sizeof(int));
    	return value;
    }

    int read_float()
    {
    	float value = -1.0f;
    	message.read( (char*)&value, sizeof(float));
    	return value;
    }

    std::string read_string()
    {
    	int size = read_int();
    	std::string str(size, ' ');

    	message.read( (char*)&str[0], size);

    	return str;
    }

    std::shared_ptr<void> read_data()
    {
    	int size = read_int();
    	std::shared_ptr<void> data( malloc(size), free );
    	message.read( (char*)data.get(), size );
    	return data;
    }

private:
	std::istringstream message;
};


// SERVER ===============================================

void execute( IScene* si, Command command )
{
	IBuffer buf((char*)command.data.get(), command.size);
	int op = buf.read_int();

	switch(op)
	{
		case EOpCode::EMessage:
		{
			si->message(buf.read_string());
		}break;
		case EOpCode::ESetAttr:
		{
			// number of attributes
			std::string handle = buf.read_string();
			int nattr = buf.read_int();

			std::vector<Attribute> attr_list;
			for( int i=0;i<nattr;++i )
			{
				// read attribute...
				Attribute attr;

				attr.m_name = buf.read_string();
				attr.m_type = buf.read_int();
				attr.m_size = buf.read_int();
				attr.m_data = buf.read_data();

				attr_list.push_back(attr);
			}

			si->setAttr( handle, &attr_list[0], attr_list.size() );
		}break;
		case EOpCode::ECreate:
		{
			std::string type = buf.read_string();
			std::string handle = buf.read_string();
			si->create( type, handle );
		}break;
		case EOpCode::ERemove:
		{
			std::string handle = buf.read_string();
			si->remove( handle );
		}break;
		default:
			//throw std::runtime_error("execute: unknown opcode");
			std::cout << "execute: unknown opcode\n";
	};
}


// CLIENT ================================================================

Command message( const std::string& text )
{
    OBuffer buf;

    buf.write( int(EOpCode::EMessage) );
    buf.write( text );

    Command cmd;
    cmd.data_alt = buf.to_string();
    cmd.size = cmd.data_alt.size();

    return cmd;
}

Command setAttr( const std::string& object, const Attribute* attr_list, int nattrs )
{
    OBuffer buf;

    buf.write( int(EOpCode::ESetAttr) );

    // serialize object handle ---
    buf.write( object );

    // serialize number of attributes ---
    buf.write( nattrs );

    // now for each attribute: expand the data buffer and serialize attribute ---
    for( int i=0;i<nattrs;++i )
    {
        // serialize attribute
        const Attribute& attr = attr_list[i];

        buf.write(attr.m_name);
        buf.write(int(attr.m_type));
        buf.write(attr.m_size);
        buf.write(attr.ptr<char>(), attr.memsize());
    }

    Command cmd;
    cmd.data_alt = buf.to_string();
    cmd.size = cmd.data_alt.size();

    return cmd;
}


